src/storage/src/upsert/types.rs

```diff
@@ -319,43 +592,62 @@ impl<O: Default> StateValue<O> {
-            *len_sum += len.wrapping_mul(diff);
-            // Truncation is fine (using `as`) as this is just a checksum
-            *checksum_sum += (seahash::hash(&*bincode_buffer) as i64).wrapping_mul(diff);
-
-            // XOR of even diffs cancel out, so we only do it if diff is odd
-            if diff.abs() % 2 == 1 {
-                if value_xor.len() < bincode_buffer.len() {
-                    value_xor.resize(bincode_buffer.len(), 0);
+        match self {
+            Self::Consolidating(Consolidating {
+                value_xor,
+                len_sum,
+                checksum_sum,
+                diff_sum,
+            }) => {
+                bincode_buffer.clear();
+                bincode_opts
+                    .serialize_into(&mut *bincode_buffer, &value)
+                    .unwrap();
+                let len = i64::try_from(bincode_buffer.len()).unwrap();
+
+                *diff_sum += diff;
+                *len_sum += len.wrapping_mul(diff);
+                // Truncation is fine (using `as`) as this is just a checksum
+                *checksum_sum += (seahash::hash(&*bincode_buffer) as i64).wrapping_mul(diff);
+
+                // XOR of even diffs cancel out, so we only do it if diff is odd
+                if diff.abs() % 2 == 1 {
+                    if value_xor.len() < bincode_buffer.len() {
+                        value_xor.resize(bincode_buffer.len(), 0);
+                    }
+                    // Note that if the new value is _smaller_ than the `value_xor`, and
+                    // the values at the end are zeroed out, we can shrink the buffer. This
+                    // is extremely sensitive code, so we don't (yet) do that.
+                    for (acc, val) in value_xor.iter_mut().zip(bincode_buffer.drain(..)) {
+                        *acc ^= val;
+                    }
                 }
-                // Note that if the new value is _smaller_ than the `value_xor`, and
-                // the values at the end are zeroed out, we can shrink the buffer. This
-                // is extremely sensitive code, so we don't (yet) do that.
-                for (acc, val) in value_xor.iter_mut().zip(bincode_buffer.drain(..)) {
-                    *acc ^= val;
+
+                // Returns whether or not the value can be deleted. This allows
+                // us to delete values in `UpsertState::consolidate_chunk` (even
+                // if they come back later), to minimize space usage.
+                diff_sum.0 == 0 && checksum_sum.0 == 0 && value_xor.iter().all(|&x| x == 0)
+            }
+            StateValue::Value(_value) => {
+                // We can turn a Value back into a Consolidating state:
+                // `std::mem::take` will leave behind a default value, which
+                // happens to be a default `Consolidating` `StateValue`.
+                let this = std::mem::take(self);
+
+                let finalized_value = this.into_finalized_value();
```

Comment:
In which case do we end up with a non-final value in here?

Reply:
We either have a finalized value or we get a `None`, if the state consolidates to a zero diff, this piece would yield a `Tombstone`: https://github.com/MaterializeInc/materialize/blob/b104dde49ff6b46e908d6b57ad6750cf96f51a47/src/storage/src/upsert/types.rs#L749